explorer/interpreter/matching_impl_set.h

```diff
@@ -0,0 +1,121 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_EXPLORER_INTERPRETER_MATCHING_IMPL_SET_H_
+#define CARBON_EXPLORER_INTERPRETER_MATCHING_IMPL_SET_H_
+
+#include <vector>
+
+#include "common/ostream.h"
+#include "explorer/ast/declaration.h"
+#include "explorer/common/nonnull.h"
+#include "explorer/interpreter/impl_scope.h"
+#include "explorer/interpreter/value.h"
+#include "llvm/ADT/DenseMap.h"
+
+namespace Carbon {
+
+// A set of impl matches that we're currently performing. Each `Match`
+// represents an attempt to match a `type as interface` query against an `impl`
+// declaration. This is used to detect and reject non-termination when impl
+// matching recursively triggers further impl matching.
+//
+// The approach we use is to count the number of times each "label" appears
+// within the type and interface, where a label is the name of a declared
+// entity such as a class or interface, or a primitive like `type` or `bool`.
+// For example, `Optional(i32*)` contains the labels for `Optional`, `i32`, and
+// `*` (built-in pointer type) once each.
+//
+// If we ever try matching the same `impl` twice, where the inner match
+// contains at least as many appearances of each label as the outer match, we
+// reject. We also reject if a query results in the exact same query being
+// performed again.
+class MatchingImplSet {
```

Comment:
```suggestion
//
// Usage pattern:
//   MatchingImplSet::Match match(&match_set, ...);
//   // Deduction creates or checks other matches.
//   match.Check(...);
//   // Deduction creates or checks other matches.
//   // match is destructed.
class MatchingImplSet {
```

One thing that might help me (still) is slight clarity on how the Match object is expected to be used, maybe particularly because it's not clear to me why you're using RAII. In particular, it's not obvious to me if you could have written this API to make Match more of an internal implementation detail. But I'm assuming a usage pattern similar to the above, which would at least explain some of the model to me.

An alternative might've been something a little more like:

```
match_set.AddMatch(&match_set, ...);
auto pop_match = OnDestroy([&]() { match_set.Pop(); }
...
match_set.CheckTop();
```

But I don't know that that's really better, so I'm not really suggesting, just commenting.

Reply:
I've added a comment to explain that this is an implementation detail of `MatchImpl`, so that readers know to look there for the usage. I've also expanded the `Match` class comment a little to cover this a bit better.

I put the match storage on the stack (in the RAII object) rather than doing something like your alternative in order to avoid heap allocation or relocation of the match data. Maybe that sort of optimization isn't so much of a concern in explorer though.